### ISSUE TITLE
problem of negative value in leadership only.

### ISSUE THREAD

**Issue by newfrenchy83** [wesnoth/wesnoth#3770]
If a unit with regular leadership ability and a unit with leadership value =- $value

by example are adjacent to tierce unit, this unit will be influenced by regular leadership ability because in attack.cpp, highest("value") is used. For this reason if two unit with

```
#define ABILITY_TERROR
    [leadership]
        id=terror
        value="(-(15 * (level - other.level) + 15))"
        cumulative=no
        name= _ "terror"
        description= _ "This unit can frighten enemy units that are next to it, making them fight worse.
Adjacent enemy units of the same or lower level will do less damage in battle. When an enemy unit of the same or lower level is adjacent and engages in combat, its attacks do 15% less damage times the difference in their levels + 15%."
        affect_self=no
        affect_allies=no
        affect_enemies=yes
        [affect_adjacent]
            [filter]
                formula="level <= other.level"
            [/filter]
        [/affect_adjacent]
    [/leadership]
#enddef
```

but two different level are adjacent to tierce unit, then the highest value (terror low efficient) will be used.

**Comment by newfrenchy83** [comment]
the fix what i want propose is in int under_leadership(const unit &u, const map_location& loc, const_attack_ptr weapon, const_attack_ptr opp_weapon)
replace return abil.highest("value").first; by int leader_up = std::max(0, abil.highest("value").first);
	int leader_down = std::min(0, abil.lowest("value").first);
	return leader_up + leader_down;
if the minimal value < 0 then tleader_down value will be the lowest value ( terror more efficient) and leader_up highest value(leadership normal more efficient) and the influence will be result of two opposite leadership

**Comment by newfrenchy83** [comment]
and evidence by image
![terror_level_2](https://user-images.githubusercontent.com/31768074/49893249-4b0dd480-fe4b-11e8-954d-e6d92133a63f.png)
![terror_level-3](https://user-images.githubusercontent.com/31768074/49893334-6973d000-fe4b-11e8-9ec0-bcb503654110.png)

the two level of terror 
![ability terror two level](https://user-images.githubusercontent.com/31768074/49893354-78f31900-fe4b-11e8-9f52-28dad0316a71.png)

and with normal leadership
![terror_and leadership_normal](https://user-images.githubusercontent.com/31768074/49893385-8e684300-fe4b-11e8-9338-4d3c65c3014f.png)

**Comment by jostephd** [comment]
> For this reason if two unit with
> 
> ```
> #define ABILITY_TERROR
> #enddef
> ```
> 
> but two different level are adjacent to tierce unit, then the highest value (terror low efficient) will be used.

@newfrenchy83 Thanks for the report. I can reproduce the problem with the following test case:

```
diff --git a/data/scenario-test.cfg b/data/scenario-test.cfg
index 50e35707ff5..4ef4e006c0e 100644
--- a/data/scenario-test.cfg
+++ b/data/scenario-test.cfg
@@ -709,9 +709,52 @@ Xu, Xu, Qxu, Qxu, Ql, Ql, Ql, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Gg, Gg, Gg, Gg, Gg
         [/filter]
         {MODIFY_UNIT (id=$second_unit.id) side 1}
     [/event]
+#define ABILITY_TERROR
+    [leadership]
+        id=terror
+        value="(-(15 * (level - other.level) + 15))"
+        cumulative=no
+        name= _ "terror"
+        description= _ "This unit can frighten enemy units that are next to it, making them fight worse.
+Adjacent enemy units of the same or lower level will do less damage in battle. When an enemy unit of the same or lower level is adjacent and engages in combat, its attacks do 15% less damage times the difference in their levels + 15%."
+        affect_self=no
+        affect_allies=no
+        affect_enemies=yes
+        [affect_adjacent]
+            [filter]
+                formula="level <= other.level"
+            [/filter]
+        [/affect_adjacent]
+    [/leadership]
+#enddef
     [event]
         name=prestart
         {VARIABLE teleports_on no}
+        {UNIT 1 (Dwarvish Thunderer) 15 8 ()}
+        {UNIT 2 Troll 14 7 (
+            [modifications]
+                [object]
+                    [effect]
+                        apply_to=new_ability
+                        [abilities]
+                            {ABILITY_TERROR}
+                        [/abilities]
+                    [/effect]
+                [/object]
+            [/modifications]
+        )}
+        {UNIT 2 (Troll Warrior) 14 8 (
+            [modifications]
+                [object]
+                    [effect]
+                        apply_to=new_ability
+                        [abilities]
+                            {ABILITY_TERROR}
+                        [/abilities]
+                    [/effect]
+                [/object]
+            [/modifications]
+        )}
     [/event]
     [label]
         x,y=20,13
```

The Thunderer's ranged attack does 10 damage (before vulnerabilities/resistances) when he's next to the lvl3 troll, 13 damage when he's next to the lvl2 troll, and 13 damage when he's adjacent to both of them. I agree that he should do 10 damage in this case.

(For future reference, @newfrenchy83, this is what I meant when I said that the best way to report an issue was _with WML code that reproduces it_. Please try to do that in the future. Runnable code is infinitely more useful than screenshots.)

I have tested your patch in #3774 and confirm fixes the issue. However, I haven't reviewed it yet. The next steps are for one of the developers to review and merge the patch. I might do that myself if I have time, or someone else might step in.

Thanks again for splitting this bug report and patch from the larger issue.

**Comment by jostephd** [comment]
Fixed by merging #3774.
